src/background.js

```diff
@@ -76,27 +74,154 @@ function getPubkey(pubkeyPatterns, url) {
+  let options = defaultOptions();
+  const head = content.split(/<\/head\s*>/i)[0];
+
+  // Parse signature metadata
+  const signatureRegex = RegExp('<meta\\s+name="signature"\\s+content="([^"]*)"\\s*>', 'gi');
+  const nameRegex = RegExp('\\s*([\\w-]+)=([^,"]*),?', 'gi');
+  let sigMatch;
+  while( (sigMatch = signatureRegex.exec(head)) !== null ) {
+    let signature = defaultSignature();
+    let nameMatch;
+    while ( (nameMatch = nameRegex.exec(sigMatch[1])) !== null ) {
+      const name = nameMatch[1].toLowerCase();
+      if (name == 'signature') {
+        // preserve whitespace to ensure all of it is stripped out
+        signature[name] = nameMatch[2];
+      } else if (name == 'allowedmethods') {
+        // split on spaces
+        signature[name] = nameMatch[2].split(" ").map(s => s.trim().toLowerCase());
+      } else {
+        signature[name] = nameMatch[2].trim().toLowerCase();
+       }
+    }
+    if (signature.type && signature.version && signature.signature)
+      options.signatures.push(signature);
+  }
+
+  return options;
+}
+
+// Strip all signatures from content for signature verification
+function stripSignatures(content, options) {
+  let newContent = content;
+  for (let signature of options.signatures)  {
+    newContent = newContent.replace(signature.signature, "");
+    signature.signature = signature.signature.trim();
+  }
+  return newContent;
+}
 
-  const shouldCheck = getPubkey(patterns, url);
+function validateSignature(content, signature, pageOptions, pubkey) {
+  const options = {
+    message: openpgp.message.fromBinary(openpgp.util.str2Uint8Array(content)),
+    signature: openpgp.signature.readArmored(signature),
+    publicKeys: openpgp.key.readArmored(pubkey).keys,
+  };
 
-  if (shouldCheck) {
-    try {
-      const pubkey = patterns[shouldCheck];
+  return openpgp.verify(options).then((verified) => {
+    return verified.signatures[0].valid;
+  });
+}
 
-      const options = {
-        message: openpgp.message.fromBinary(openpgp.util.str2Uint8Array(content)),
-        signature: openpgp.signature.readArmored(signature),
-        publicKeys: openpgp.key.readArmored(pubkey).keys,
-      };
+function validateSignatures(content, options, pubKey, method) {
+  let last = Promise.resolve(false);
+  for (let signature of options.signatures) {
+    let promise = null;
+    if (signature.allowedmethods.includes(method.toLowerCase())) {
+      if (signature.type == 'pgp') {
+        promise = validateSignature(content, signature.signature, options, pubKey);
+      } else if (signature.type == 'pgpMinimized') {
+        if (matchVersions(signature.version, '1.0.0')) {
+          const signedContent = minimize_1_0(content);
+          promise =  validateSignature(signedContent, signature.signature, options, pubKey);
+        }
+      }
+    }
+    // Chain promises returning when the first one is true
+    if (promise) {
+      last = last.then(verified => {
+        if (verified)
```

Comment:
promise.race only returns one value, the first.  Promise.all only returns when all the values are completed.  This code goes through the signatures one at a time, from first to last and returns true when one of them returns true.  Since verification is CPU and not IO, I don't know how much parallelism we actually get.

Reply:
Updated the code to hopefully be cleaner